csharp/src/Microsoft.ML.OnnxRuntime/Training/NativeTrainingMethods.cs

```diff
@@ -0,0 +1,256 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.ML.OnnxRuntime
+    {
+        // OrtTrainingApi  (onnxruntime_training_c_api.cc)
+        [StructLayout(LayoutKind.Sequential)]
+        public struct OrtTrainingApi
+        {
+            public IntPtr LoadCheckpoint;
+            public IntPtr SaveCheckpoint;
+            public IntPtr CreateTrainingSession;
+            public IntPtr TrainingSessionGetTrainModelOutputCount;
+            public IntPtr TrainingSessionGetEvalModelOutputCount;
+            public IntPtr TrainingSessionGetTrainModelOutputName;
+            public IntPtr TrainingSessionGetEvalModelOutputName;
+            public IntPtr ResetGrad;
+            public IntPtr TrainStep;
+            public IntPtr EvalStep;
+            public IntPtr SetLearningRate;
+            public IntPtr GetLearningRate;
+            public IntPtr OptimizerStep;
+            public IntPtr RegisterLinearLRScheduler;
+            public IntPtr SchedulerStep;
+            public IntPtr ReleaseTrainingSession;
+            public IntPtr ReleaseCheckpointState;
+        }
+
+        internal static class NativeTrainingMethods
+        {
+            static OrtApi api_;
+            static OrtTrainingApi trainingApi_;
+            static IntPtr trainingApiPtr;
+
+            [UnmanagedFunctionPointer(CallingConvention.Winapi)]
+            public delegate ref OrtApi DOrtGetApi(UInt32 version);
+
+            [UnmanagedFunctionPointer(CallingConvention.Winapi)]
+            public delegate IntPtr /* OrtTrainingApi* */ DOrtGetTrainingApi(UInt32 version);
+            public static DOrtGetTrainingApi OrtGetTrainingApi;
+
+        static NativeTrainingMethods()
+            {
+                DOrtGetApi OrtGetApi = (DOrtGetApi)Marshal.GetDelegateForFunctionPointer(NativeMethods.OrtGetApiBase().GetApi, typeof(DOrtGetApi));
+
+                // TODO: Make this save the pointer, and not copy the whole structure across
+                api_ = (OrtApi)OrtGetApi(4 /*ORT_API_VERSION*/);
```

Comment:
>  api_ = (OrtApi)OrtGetApi(4 /*ORT_API_VERSION*/); [](http://example.com/codeflow?start=14&length=50)

Can you build an inference only onnxruntime.dll and see what happens if you try to run your changes?

Reply:
Yes I have done that testing